core/src/message/context.rs

```diff
@@ -241,31 +201,47 @@ impl ContextStore {
 }
 
 /// Context of currently processing incoming message.
-#[derive(Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+#[derive(Debug)]
 pub struct MessageContext {
     kind: DispatchKind,
     current: IncomingMessage,
     outcome: ContextOutcome,
     store: ContextStore,
     settings: ContextSettings,
+    outgoing_bytes_counter: u32,
 }
 
 impl MessageContext {
-    /// Create new MessageContext with given ContextSettings.
+    /// Create new message context.
+    /// Returns `None` if outgoing messages bytes limit exceeded.
     pub fn new(
         dispatch: IncomingDispatch,
         program_id: ProgramId,
         settings: ContextSettings,
-    ) -> Self {
+    ) -> Option<Self> {
         let (kind, message, store) = dispatch.into_parts();
 
-        Self {
+        let outgoing_bytes_counter = match &store {
+            Some(store) => {
+                let mut counter = 0u32;
+                for payload in store.outgoing.values().filter_map(|x| x.as_ref()) {
+                    counter = counter.checked_add(payload.len_u32())?;
+                }
+                counter
+            }
+            None => 0,
+        };
+
+        (outgoing_bytes_counter <= settings.outgoing_bytes_limit).then_some(())?;
+
+        Some(Self {
             kind,
             outcome: ContextOutcome::new(program_id, message.source(), message.id()),
             current: message,
             store: store.unwrap_or_default(),
             settings,
-        }
+            outgoing_bytes_counter,
```

Comment:
> it saves us both from reallocs on journal notes creation + stored context creation

It does not save as from allocator overflow. You can make a lot of send_push without send_commit in first execution before wait. After wake we should takes in account these bytes because they take runtime memory.

Reply:
Got you point, overlooking it